`FontSizeAdjust`'s `metric` member is uninitialized

obyknovenius opened the pull request:
#### 493fe8cbc12cfd94a262852d3c0d0fad3ac8ffb3
<pre>
`FontSizeAdjust`&apos;s `metric` member is uninitialized
<a href="https://bugs.webkit.org/show_bug.cgi?id=256034">https://bugs.webkit.org/show_bug.cgi?id=256034</a>

Reviewed by Myles C. Maxfield.

Because of it two otherwise identical `RenderStyle`s
will be considered different and can cause full relayout
in some circumstances.

* Source/WebCore/platform/graphics/FontSizeAdjust.h:

Canonical link: <a href="https://commits.webkit.org/264195@main">https://commits.webkit.org/264195@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40780e923f9ec9c54b6450a0ed933b972968c7e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6887 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7117 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8486 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7134 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6892 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7056 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10035 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7009 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7670 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6340 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8586 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5052 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6251 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14047 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6739 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6338 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9152 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6819 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5584 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6195 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6165 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1657 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10380 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6573 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -61,8 +61,10 @@ struct FontSizeAdjust {
+    };
+    
+    Metric metric { };
+    Markable<float, FloatMarkableTraits> value { };
```
Isn't that exactly the same with 4 more characters ? (aka calling the default constructor) 

`Foo foo` and `Foo foo { }` ?

obyknovenius replied:
I thought this too but seems like it is not. At least on Mac. Maybe because of `clang`.

mdubet replied:
Can we write a test case which demonstrate it ?

obyknovenius replied:
I've added an API test.

obyknovenius replied:
I've updated the test. It should now better illustrate the problem.

shivamidow commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -61,8 +61,10 @@ struct FontSizeAdjust {
-    Markable<float, FloatMarkableTraits> value;
+    };
+    
+    Metric metric { };
```
Let's be explicit about the initial value once we decide to initialize it. `Metric metric { Metric::ExHeight };`

obyknovenius replied:
Done

shivamidow commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -61,8 +61,10 @@ struct FontSizeAdjust {
+    };
+    
+    Metric metric { };
+    Markable<float, FloatMarkableTraits> value { };
```
Not sure about this initialization. Markable should be initialized with emptyValue() though, does the value also cause a problem?

obyknovenius replied:
I've reverted this change.

shivamidow commented on Tools/TestWebKitAPI/Tests/WebCore/FontSizeAdjust.cpp:
```diff
@@ -0,0 +1,41 @@
+
+TEST(FontSizeAdjustTest, Metric)
+{
+    FontSizeAdjust fontSizeAdjust;
```
The FontSizeAdjust struct is initialized via the list initialization when it is created. There is no declaration or use like this in the current code IIRC. Thus, we got no style issue.
By the way, this test just checks if an uninitialized struct has a garbage value or not. As others said, it might be better to have a web test showing a problem in practice instead...

obyknovenius replied:
I've updated the test. It should now better illustrate the problem. The new test basically simulates this logic:
https://github.com/WebKit/WebKit/blob/3b9112ff64a6aa1c0daf62390d1646bcc36b488b/Source/WebCore/dom/Document.cpp#L2142-L2145
It is hard to write a web test for it because the issue is that the style update will be propagated starting from `renderView` down the tree. But this can hide other problems like #12312 which is now only reproducible on Linux, probably because of using `gcc` instead of `clang`.

shivamidow reviewed (commented):
/* Unofficial Review */
I commented below. The uninitialized FontSizeAdjust might not cause a practical problem, but I like things clear than unclear. =)

vitorroriz commented on LayoutTests/fast/css/no-style-change-no-repaint.html:
```diff
@@ -0,0 +1,19 @@
+        document.getElementById('layers').textContent = internals.layerTreeAsText(document, internals.LAYER_TREE_INCLUDES_REPAINT_RECTS);
+}
+</script>
+<pre id="layers"></pre>
```
*disclaimer: Unofficial review (I'm not a reviewer)*

Can you transform this test in a reftest? The -ref will be pretty similar with the test itself, but without the `<script> `part and with hard-coded `textContent` written into the `id="layers"` element. In that way, the -expected file don't need to be a render dump.
reference: https://web-platform-tests.org/writing-tests/reftests.html

obyknovenius replied:
Done.

vitorroriz commented on LayoutTests/fast/css/no-style-change-no-repaint-expected.txt:
```diff
@@ -0,0 +1,7 @@
+  RenderBlock {HTML} at (0,0) size 800x600
+    RenderBody {BODY} at (8,8) size 784x584
+      RenderBlock {DIV} at (0,0) size 784x0
+      RenderBlock {PRE} at (0,0) size 784x0
```
*disclaimer: Unofficial review (I'm not a reviewer)*

This could be a simple copy of a -ref file, instead of a render dump.
See my other comment on the actual test.

obyknovenius replied:
Done.

vitorroriz commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -288,7 +288,7 @@ class RenderStyle {
     static RenderStyle& defaultStyle();
 
-    static RenderStyle create();
+    WEBCORE_EXPORT static RenderStyle create();
```
*disclaimer: I'm not an official reviewer*.

I'm not sure if we should to do that. You can probably remove your .cpp test since you have introduced a web test.

brentfulgham replied:
I think this is okay -- it's being exposed for a TestWebKitAPI purpose. We don't really have a good mechanism to expose things only for API tests/Unit Tests, which would be a good project for someone.

vitorroriz commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1159,7 +1159,7 @@ class RenderStyle {
     void setTableLayout(TableLayoutType v) { m_nonInheritedFlags.tableLayout = static_cast<unsigned>(v); }
 
-    bool setFontDescription(FontCascadeDescription&&);
+    WEBCORE_EXPORT bool setFontDescription(FontCascadeDescription&&);
```
*disclaimer: I'm not an official reviewer*.

ditto.

vitorroriz commented on Source/WebCore/style/StyleChange.h:
```diff
@@ -40,7 +40,7 @@ enum class Change : uint8_t {
 };
 
-Change determineChange(const RenderStyle&, const RenderStyle&);
+WEBCORE_EXPORT Change determineChange(const RenderStyle&, const RenderStyle&);
```
*disclaimer: I'm not an official reviewer*.

ditto.

brentfulgham reviewed (commented):
I think this change looks good, but I'm less confident in the web test. It seems to work fine for modern WebKit, but the WebKitLegacy test bot is unhappy. Is there a reason this behavior should be different on legacy WebKit? At the very least, we should skip the test on that port to avoid slowing down the bots.

obyknovenius committed:
`FontSizeAdjust`'s `metric` member is uninitialized
https://bugs.webkit.org/show_bug.cgi?id=256034

Reviewed by Myles C. Maxfield.

Because of it two otherwise identical `RenderStyle`s
will be considered different and can cause full relayout
in some circumstances.

* Source/WebCore/platform/graphics/FontSizeAdjust.h:

Canonical link: https://commits.webkit.org/264195@main
